File: app/src/main/java/me/zhanghai/android/files/provider/common/AbstractWatchKey.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.common;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import java8.nio.file.Path;
import java8.nio.file.StandardWatchEventKinds;
import java8.nio.file.WatchEvent;
import java8.nio.file.WatchKey;
import me.zhanghai.android.files.util.CollectionUtils;

public abstract class AbstractWatchKey implements WatchKey {

    private static final int MAX_PENDING_EVENTS = 256;

    @NonNull
    private final AbstractWatchService mWatchService;
    @NonNull
    private final Path mPath;

    private boolean mSignaled;
    private List<Event<?>> mEvents = new ArrayList<>();

    @NonNull
    protected final Object mLock = new Object();

    public AbstractWatchKey(@NonNull AbstractWatchService watchService, @NonNull Path path) {
        mWatchService = Objects.requireNonNull(watchService);
        mPath = Objects.requireNonNull(path);
    }

    public <T> void addEvent(@NonNull WatchEvent.Kind<T> kind, @Nullable T context) {
        Objects.requireNonNull(kind);
        synchronized (mLock) {
            if (!mEvents.isEmpty()) {
                Event<?> lastEvent = CollectionUtils.last(mEvents);
                if (lastEvent.kind() == StandardWatchEventKinds.OVERFLOW
                        || (lastEvent.kind() == kind && Objects.equals(lastEvent.context(),
                        context))) {
                    lastEvent.repeat();
                    return;
                }
            }
            if (kind == StandardWatchEventKinds.OVERFLOW || mEvents.size() >= MAX_PENDING_EVENTS) {
                mEvents.clear();
                mEvents.add(new Event<>(StandardWatchEventKinds.OVERFLOW, null));
                signal();
                return;
            }
            mEvents.add(new Event<>(kind, context));
            signal();
        }
    }

    public void signal() {
        synchronized (mLock) {
            if (!mSignaled) {
                mSignaled = true;
                mWatchService.enqueue(this);
            }
        }
    }

    @NonNull
    @Override
    public List<WatchEvent<?>> pollEvents() {
        synchronized (mLock) {
            List<Event<?>> events = mEvents;
            mEvents = new ArrayList<>();
            //noinspection unchecked
            return (List<WatchEvent<?>>) (List<?>) events;
        }
    }

    @Override
    public boolean reset() {
        synchronized (mLock) {
            boolean valid = isValid();
            if (valid && mSignaled) {
                if (mEvents.isEmpty()) {
                    mSignaled = false;
                } else {
                    mWatchService.enqueue(this);
                }
            }
            return valid;
        }
    }

    @NonNull
    @Override
    public Path watchable() {
        return mPath;
    }

    @NonNull
    protected AbstractWatchService getWatchService() {
        return mWatchService;
    }


    private static class Event<T> implements WatchEvent<T> {

        @NonNull
        private final WatchEvent.Kind<T> mKind;
        @Nullable
        private final T mContext;

        private int mCount = 1;

        @NonNull
        private final Object mLock = new Object();

        Event(@NonNull WatchEvent.Kind<T> type, @Nullable T context) {
            mKind = Objects.requireNonNull(type);
            mContext = context;
        }

        @NonNull
        @Override
        public WatchEvent.Kind<T> kind() {
            return mKind;
        }

        @Nullable
        @Override
        public T context() {
            return mContext;
        }

        @Override
        public int count() {
            return mCount;
        }

        void repeat() {
            synchronized (mLock) {
                ++mCount;
            }
        }
    }
}

File: app/src/main/java/me/zhanghai/android/files/provider/common/AbstractWatchKey.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.common

import java8.nio.file.Path
import java8.nio.file.StandardWatchEventKinds
import java8.nio.file.WatchEvent
import java8.nio.file.WatchKey

abstract class AbstractWatchKey<K : AbstractWatchKey<K, P>, P : Path>(
    private val watchService: AbstractWatchService<K>,
    private val path: P
) : WatchKey {
    private var isValid = true

    private var isSignaled = false

    private var events = mutableListOf<Event<*>>()

    protected val lock = Any()

    override fun isValid(): Boolean {
        synchronized(lock) { return isValid }
    }

    fun setInvalid() {
        synchronized(lock) { isValid = false }
    }

    fun <T> addEvent(kind: WatchEvent.Kind<T>, context: T?) {
        synchronized(lock) {
            if (events.isNotEmpty()) {
                val lastEvent = events.last()
                if (lastEvent.kind() == StandardWatchEventKinds.OVERFLOW
                    || (lastEvent.kind() == kind && lastEvent.context() == context)) {
                    lastEvent.repeat()
                    return
                }
            }
            if (kind === StandardWatchEventKinds.OVERFLOW || events.size >= MAX_PENDING_EVENTS) {
                events.clear()
                events.add(Event(StandardWatchEventKinds.OVERFLOW, null))
                signal()
                return
            }
            events.add(Event(kind, context))
            signal()
        }
    }

    fun signal() {
        synchronized(lock) {
            if (!isSignaled) {
                isSignaled = true
                @Suppress("UNCHECKED_CAST")
                watchService.enqueue(this as K)
            }
        }
    }

    override fun pollEvents(): List<WatchEvent<*>> {
        synchronized(lock) {
            val events = events
            this.events = mutableListOf()
            return events
        }
    }

    override fun reset(): Boolean {
        synchronized(lock) {
            val isValid = isValid
            if (isValid && isSignaled) {
                if (events.isEmpty()) {
                    isSignaled = false
                } else {
                    @Suppress("UNCHECKED_CAST")
                    watchService.enqueue(this as K)
                }
            }
            return isValid
        }
    }

    override fun cancel() {
        synchronized(lock) {
            if (isValid) {
                @Suppress("UNCHECKED_CAST")
                watchService.cancel(this as K)
            }
        }
    }

    override fun watchable(): P = path

    private class Event<T> constructor(
        private val kind: WatchEvent.Kind<T>,
        private val context: T?
    ) : WatchEvent<T> {
        private var count = 1

        private val lock = Any()

        override fun kind(): WatchEvent.Kind<T> = kind

        override fun context(): T? = context

        override fun count(): Int = count

        fun repeat() {
            synchronized(lock) { ++count }
        }
    }

    companion object {
        private const val MAX_PENDING_EVENTS = 256
    }
}
